### ISSUE TITLE
Caption's interaction with width-reduced listings

### ISSUE THREAD

**Issue by Piraty** [axelsommerfeldt/caption#1]
### Setup
- linux
- texlive2017 (up to date as of 2017-10-02).

### Background
i tend to use floating listings, like you can observe below in the MWE.
I saw unexpected behaviour using the caption package, in the following condition:

To reduce the width of those listings, `xleftmargin` and `xrightmargin` are set to equal values to mimic centering.

### Issue
Depending on the resulting width of the listing (and/or the length of the caption?), the listing's caption does not respect the justification and is not centered.
Simply removing the `\usepackage{caption}` (and `\captionsetup` therefore too of course) puts the caption right in the center where it is intended to be.

![listing-centering](https://user-images.githubusercontent.com/23238955/31120688-ab2f9136-a835-11e7-8444-959f98e4d287.png)

### More
Can you confirm?
Can you tell a workaround?

I found [some info from you](https://tex.stackexchange.com/a/88206/85772) that did not directly relate to width-reduced listings, so i poke you with a new issue ;-)

I also opened a question on [tex.stackexchange.com](https://tex.stackexchange.com/q/394247/85772)

### MWE
```
\documentclass[]{scrbook}

\usepackage{xcolor}
\usepackage{listings}
\usepackage{caption}

\captionsetup{
	justification = centering, % justified, centering, RaggedRight
}

\begin{document}
\begin{lstlisting}[
	frame=shadowbox %helper, to see listing's dim
	,captionpos=b
	%,float %float is irrelevant to the issue
	%center the listing
		,xleftmargin=.20\textwidth
		,xrightmargin=.20\textwidth
	,label=code
	,caption={This caption is not centered at all}
	,language=Delphi]
i := 1;
\end{lstlisting}
	
\end{document}

```

**Comment by u-fischer** [comment]
This just came up again (https://tex.stackexchange.com/questions/428101/package-listings-set-caption-position-posotion-top-works-fine-but-option-bott) and I investigated. Imho a change like this should remedy the problem:

```
\documentclass{article}

\usepackage{color}

\usepackage{listings}


\usepackage[justification=centering,singlelinecheck=false]{caption}
\makeatletter

 \def\lst@MakeCaption#1{% #1 is `t' or `b'
    \begingroup 
      \caption@setposition{#1}%
      \caption@iftop{%
        \ifdim\hsize>\linewidth
          \hsize\linewidth
        \fi
        \@tempdima\belowcaptionskip
        \belowcaptionskip\abovecaptionskip
        \abovecaptionskip\@tempdima}
        {\ifdim\hsize>\dimexpr \linewidth + \lst@xleftmargin + \lst@xrightmargin\relax
          \hsize\dimexpr \linewidth + \lst@xleftmargin + \lst@xrightmargin\relax
        \fi}%
      \caption@setup{rule=0}%
      \caption@setoptions{lstlisting}%
      \caption@setautoposition{#1}%
      \caption@begin{lstlisting}%
        \caption@ORI@lst@MakeCaption{#1}%
      \caption@end
    \endgroup}%
\makeatother

\lstdefinestyle{SC}{
    language=erlang,
    xleftmargin=.2\textwidth,
    xrightmargin=.2\textwidth
}

\lstset{
    numbers=left,
    numberstyle=\tiny,
    basicstyle=\scriptsize,
    showstringspaces=false,
    showspaces=false,
    commentstyle=\color{blue},
    keywordstyle=\bfseries\color{purple},
    captionpos=t,
    frame=lines
}

\begin{document}

\begin{lstlisting}[style=SC, caption={test},captionpos=t]
xxxxxxxxxxxxxxxxxxxxx\\
xxxxxxxxxxxxxxxxxxxxxxxx
\end{lstlisting}

\begin{lstlisting}[style=SC, caption={test},captionpos=b]
% Stop all containers
docker stop $(docker ps -a -q)
% Delete all containers
docker rm $(docker ps -a -q)
% Delete all images
docker rmi $(docker images -q)
\end{lstlisting}

\end{document}
```

**Comment by davidwowa** [comment]
But if work is bigger than comming error `pdflatex.exe> ! TeX capacity exceeded, sorry [grouping levels=255].
pdflatex.exe> \caption@@ifinlist #1#2->\begingroup 
pdflatex.exe>                                      \def \@tempa ##1,#1,##2\@nil {\endgroup...
pdflatex.exe> l.9 Tendermint}, captionpos=b]{code/tender_att.sh}` and pdf file is not produced. :-(

And I add code examples as 
`
\lstinputlisting[style=MyShell, caption={Tendermint Attack}, captionpos=b]{code/tender_att.sh}
`

Here is my `root.tex` file:

`
\listfiles%
\documentclass[a4paper, 11pt, twoside]{scrreprt}
\usepackage[automark,headsepline,ilines,komastyle,autooneside]{scrpage2} 

\usepackage{geometry}
\geometry{head=4cm, headsep=15mm, height=232mm, bottom=25mm, left=3cm,
right=3cm}

\usepackage[ngerman]{babel}
\usepackage[T1]{fontenc} 
\usepackage[utf8]{inputenc}

%% for tikz
\usepackage{tikz}
\usetikzlibrary{shapes, positioning, arrows, automata}
\usepackage{smartdiagram}
 
\usepackage{subfig}
\usepackage[autostyle=true, german=quotes]{csquotes}
\usepackage{svg}

%% Auf Schriftart Palatino umschalten
\usepackage{mathpazo}
\usepackage[scaled=.95]{helvet}

%% for lines
%\usepackage{booktabs}

\usepackage{amsmath}

%% for acronyms
\usepackage[printonlyused, withpage, smaller]{acronym}
\usepackage{relsize}

\usepackage{hyperref}
\usepackage{graphicx}
\usepackage{makeidx}
\usepackage{color}
%% Damit die Bilder und Tabellen nicht auf nächste Seite weglaufen
\usepackage{float}
%% Tables
\usepackage{longtable}
\usepackage{tablefootnote}
\usepackage{threeparttable}

\usepackage[square]{natbib}

%% \usepackage{courier}
\usepackage{listings}
%\usepackage[justification=centering]{caption}
%\usepackage{listings-golang}
%% \usepackage{alltt}
\usepackage{amssymb}
\usepackage{latexsym}
%% \usepackage{rotating}
%% Hiermit kann man ganze Seite drehen, mit begin{landscape}
%% \usepackage{pdflscape}
%% \usepackage{lscape}
\usepackage{pifont}

%% for labeling in itemizing
\usepackage{scrextend}

\usepackage[justification=centering,singlelinecheck=false]{caption}
\makeatletter

 \def\lst@MakeCaption#1{% #1 is `t' or `b'
    \begingroup 
      \caption@setposition{#1}%
      \caption@iftop{%
        \ifdim\hsize>\linewidth
          \hsize\linewidth
        \fi
        \@tempdima\belowcaptionskip
        \belowcaptionskip\abovecaptionskip
        \abovecaptionskip\@tempdima}
        {\ifdim\hsize>\dimexpr \linewidth + \lst@xleftmargin + \lst@xrightmargin\relax
          \hsize\dimexpr \linewidth + \lst@xleftmargin + \lst@xrightmargin\relax
        \fi}%
      \caption@setup{rule=0}%
      \caption@setoptions{lstlisting}%
      \caption@setautoposition{#1}%
      \caption@begin{lstlisting}%
        \caption@ORI@lst@MakeCaption{#1}%
      \caption@end
    \endgroup}%
\makeatother

\lstdefinestyle{MyShell}{language=bash}

\lstdefinestyle{SC}{
	language=erlang,
	xleftmargin=.2\textwidth,
	xrightmargin=.2\textwidth
}

\lstset{ 
	numbers=left,
	numberstyle=\tiny,
	basicstyle=\scriptsize,
	showstringspaces=false,
	showspaces=false,
	commentstyle=\color{blue},
	keywordstyle=\bfseries\color{purple},
	captionpos=b,
	frame=lines
}

\makeindex

\bibliographystyle{alphadin}

% damit jede Seite einen Strich oben hat wichtig!
% Funktioniert noch nicht, in Kapitelseiten noch kein Seitenzahl
% \ofoot{\pagemark}
\setheadtopline{2pt}
\setheadsepline{.4pt} 
\renewcommand{\thechapter}{\arabic{chapter}}
\renewcommand{\chapterpagestyle}{scrheadings}
\pagestyle{scrheadings}

\begin{document}

\pagenumbering{roman}
\include{trailer}

\pagenumbering{arabic} 
\setcounter{page}{1}

%% Kapiteln
\include{chapter/introduction}
\include{chapter/problem_description}
\include{chapter/fundamentals}
\include{chapter/arch_analysis}
\include{chapter/system_description}
\include{chapter/conclusion}

\begin{appendix}
\include{appendix/bitcoin/bitcoin}
\include{appendix/profs/profs}
\include{appendix/tendermint/tendermint}
\include{appendix/attacks}
\end{appendix}

% Literaturliste soll im Inhaltsverzeichnis auftauchen
\addcontentsline{toc}{chapter}{Literatur}
\bibliography{content/contents}

\listoftable
\listoffigures

% Index soll Stichwortverzeichnis heissen
\newpage
\renewcommand{\indexname}{Stichwortverzeichnis}
	 
% Stichwortverzeichnis soll im Inhaltsverzeichnis auftauchen
\addcontentsline{toc}{chapter}{Stichwortverzeichnis}

% Stichwortverzeichnis endgueltig anzeigen
\printindex

\section*{Abkürzungsverzeichnis}\label{sec:abk} 
\begin{acronym}
 \acro{CU}{Calcualte Unit}
\end{acronym}

\end{document}
`

**Comment by axelsommerfeldt** [comment]
Try applying the patch by

    \makeatletter
    \AtBeginDocument{%  <== This line added
    \def\lst@MakeCaption#1{% #1 is t' orb'
    ...
    \endgroup}%
    }%  <== This line added
    \makeatother

instead.

**Comment by axelsommerfeldt** [comment]
Fixed with 0a32b2b8d4f814c58da1969ac7f38c7f9ec8f183. Thanks a lot for the bugfix, Ulrike.

**Comment by axelsommerfeldt** [comment]
Prepared for CTAN release with commit a5a0e2b65478ed4fb8d1239cd933264479a1d0b2
